### PR TITLE
Add ClusterRbacConfig controller, work queue, and full sync

### PR DESCRIPTION
**Pull request by mcieplak**
@prabushyam

### CODE REVIEW

File: pkg/istio/onboarding/controller.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+		return c.createClusterRbacConfig(serviceList)
+	}
+
+	clusterRbacConfig, ok := config.Spec.(*v1alpha1.RbacConfig)
+	if !ok {
+		return errors.New("Could not cast to ClusterRbacConfig")
+	}
+
+	if clusterRbacConfig.Inclusion == nil {
+		return errors.New("ClusterRbacConfig inclusion service list is empty")
```
Comment (prabushyam) [inline pkg/istio/onboarding/controller.go]: is this an error condition, in the case where the crc resource exists, but not the `Inclusion` field, this would return an error... if there are services to be onboarded, we could add the inclusion field and then process the service list

File: pkg/istio/onboarding/controller.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+			Spec:       clusterRbacConfig,
+		})
+		return err
+	}
+
+	log.Println("Sync state is current, no changes needed...")
+	return nil
+}
+
+// compareServices compares the two arrays and returns the difference
```
Comment (prabushyam) [inline pkg/istio/onboarding/controller.go]: can we update the comment to explicitly indicate that this func returns the items in first list that are not in the second list

File: pkg/istio/onboarding/controller.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+		if _, exists := serviceMapB[item]; !exists {
+			serviceListDiff = append(serviceListDiff, item)
+		}
+	}
+
+	return serviceListDiff
+}
+
+// remove removes an element from an array at the given index
+func removeIndexElement(serviceList []string, indexToRemove int) []string {
```
Comment (prabushyam) [inline pkg/istio/onboarding/controller.go]: 💯  nice optimization

File: pkg/istio/onboarding/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+			name:             "test processing a service object",
+			inputObject:      onboardedService,
+			expectedQueueLen: 1,
+			expectedKeyName:  "test-namespace/onboarded-service",
+		},
+		{
+			name:             "test processing a non service object",
+			inputObject:      nil,
+			expectedQueueLen: 0,
+		},
```
Comment (prabushyam) [inline pkg/istio/onboarding/controller_test.go]: could we add a notOnboardedService test case being added to queue as well?

File: pkg/istio/onboarding/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+func TestAddService(t *testing.T) {
+	c := &Controller{}
+
+	tests := []struct {
+		name              string
+		inputServices     []string
+		clusterRbacConfig *v1alpha1.RbacConfig
+		expectedArray     []string
+	}{
+		{
```
Comment (prabushyam) [inline pkg/istio/onboarding/controller_test.go]: could we add a nil `clusterRbacConfig:` test case?

Comment (prabushyam) [reply pkg/istio/onboarding/controller_test.go]: also, an empty `inputServices` list?

File: pkg/istio/onboarding/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+			name:              "test deleting service from ClusterRbacConfig",
+			clusterRbacConfig: createClusterRbacConfig(onboardedServiceName),
+			inputArray:        []string{onboardedServiceName},
+			expectedArray:     []string{},
+		},
+		{
+			name:              "test deleting existing service from ClusterRbacConfig",
+			clusterRbacConfig: createClusterRbacConfig(onboardedServiceName, existingServiceName),
+			inputArray:        []string{onboardedServiceName},
+			expectedArray:     []string{existingServiceName},
```
Comment (prabushyam) [inline pkg/istio/onboarding/controller_test.go]: could we add test cases for empty inputArray, and a servce that doesn't exist on the clusterRbacConfig

File: pkg/istio/onboarding/controller.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+)
+
+type Controller struct {
+	store                model.ConfigStoreCache
+	dnsSuffix            string
+	serviceIndexInformer cache.SharedIndexInformer
+	queue                workqueue.RateLimitingInterface
+}
+
+type metaNamespaceKeyFunc func(obj interface{}) (string, error)
```
[the remark this answers is not in the document]
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Ah nice! Updated it.

File: pkg/istio/onboarding/controller.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+		dnsSuffix:            dnsSuffix,
+		serviceIndexInformer: serviceIndexInformer,
+		queue:                queue,
+	}
+
+	serviceIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.processEvent(cache.MetaNamespaceKeyFunc, obj)
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
```
[the remark this answers is not in the document]
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: New surprisingly isn't a keyword but new() is a function which is supported which initializes an object with empty values. I have renamed it to avoid confusion.

File: pkg/istio/onboarding/controller.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.sync()
+	if err != nil {
+		log.Println("Error syncing cluster rbac config:", err)
```
[the remark this answers is not in the document]
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Yep, updated!

File: pkg/istio/onboarding/controller.go [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+		if c.queue.NumRequeues(key) < numRetries {
+			c.queue.AddRateLimited(key)
```
[the remark this answers is not in the document]
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Sure, added it.

File: pkg/istio/onboarding/controller.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+// addService will add a service to the ClusterRbacConfig object
+func (c *Controller) addServices(services []string, clusterRbacConfig *v1alpha1.RbacConfig) {
+	for _, service := range services {
+		clusterRbacConfig.Inclusion.Services = append(clusterRbacConfig.Inclusion.Services, service)
```
[the remark this answers is not in the document]
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Yes, good point. Since we are removing the check from the sync function, I have added it here now.

File: pkg/istio/onboarding/controller.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+		},
+	}
+
+	_, err := c.store.Create(config)
+	return err
+}
+
+// getServiceList extracts all services from the indexer with the authz
+// annotation set to true.
+func (c *Controller) getServiceList() []string {
```
[the remark this answers is not in the document]
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Renamed to getOnboardedServiceList.

File: pkg/istio/onboarding/controller.go (hunk 1 above)
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: That's a good point, it could still exist and we should add service to the object in those cases. I have removed this if statement for those use cases.

File: pkg/istio/onboarding/controller.go (hunk 2 above)
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Updated!

File: pkg/istio/onboarding/controller.go (hunk 3 above)
Comment (mcieplak) [reply pkg/istio/onboarding/controller.go]: Thanks to stack overflow :)

File: pkg/istio/onboarding/controller_test.go (hunk 4 above)
Comment (mcieplak) [reply pkg/istio/onboarding/controller_test.go]: Added the test case.

File: pkg/istio/onboarding/controller_test.go (hunk 5 above)
Comment (mcieplak) [reply pkg/istio/onboarding/controller_test.go]: Sure, added the tests.

File: pkg/istio/onboarding/controller_test.go (hunk 6 above)
Comment (mcieplak) [reply pkg/istio/onboarding/controller_test.go]: Updated.

**Comment by mcieplak** [comment]
@prabushyam I have updated the PR, please review again. Thanks!
